File: api/user_profile/user_profile_summarization_agent.py
import logging
from typing import List, Optional
from langchain_anthropic import ChatAnthropic
from pydantic import BaseModel
from user_profile.prompt.summarize_user_profile import prompt
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers import StrOutputParser


class UserProfileSummarizationResult(BaseModel):
	job_title: Optional[str]
	locations: Optional[List[str]]
	skills: Optional[List[str]]
	experience: Optional[List[str]]
	education: Optional[List[str]]
	job_contents: Optional[List[str]]

class UserProfileSummarizationAgent:
	def __init__(self, model):
		self._model = model
		self._prompt = ChatPromptTemplate.from_messages([
			("system", prompt["system"]),
			("user", "{user_profile}")
		])
		self._chain = self._prompt | self._model | StrOutputParser()

	def run(self, user_profile: str, retries=3):
		logging.debug("Start summarizing user profile...")
		retry = 0
		while retry < retries:
			try:
				output: str = self._chain.invoke({
					"user_profile": user_profile
				})
				print(">>>>", output)
				starting_index = output.find("{")
				output = output[starting_index:]
				res = UserProfileSummarizationResult.model_validate_json(output)
				if res:
					return res
			except Exception as e:
				logging.debug(f"Output: {output}")
				logging.error(f"Failed to summarize user profile: {e}, retrying...")
			retry += 1

		logging.error(f"Failed to summarize user profile job after {retry} retries.")
		return None

File: api/src/agents/user_profile/user_profile_summarization_agent.ts
import { Expose, plainToInstance } from 'class-transformer'
import {
  IsOptional,
  IsString,
  IsArray,
  validateOrReject,
} from 'class-validator'
import { ChatPromptTemplate } from '@langchain/core/prompts'
import { RunnableLike, Runnable } from '@langchain/core/runnables'
import { prompt } from './prompt/user_profile_summarization_few_shot'
import 'reflect-metadata'
import { StringOutputParser } from '@langchain/core/output_parsers'
import { Logger } from '@nestjs/common'

export class UserProfileSummarizationResult {
  @Expose({ name: 'job_title' })
  @IsOptional()
  @IsString()
  jobTitle?: string

  @Expose({ name: 'locations' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  locations?: string[]

  @Expose({ name: 'skills' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  skills?: string[]

  @Expose({ name: 'experiences' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  experiences?: string[]

  @Expose({ name: 'education' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  education?: string[]

  @Expose({ name: 'job_contents' })
  @IsOptional()
  @IsArray()
  @IsString({ each: true })
  jobContents?: string[]
}

export class UserProfileSummarizationAgent {
  private chain: Runnable

  constructor(model: RunnableLike) {
    const promptTemplate = ChatPromptTemplate.fromMessages([
      ['system', prompt.system],
      ['user', '{profile}'],
    ])
    this.chain = promptTemplate.pipe(model).pipe(new StringOutputParser())
  }

  async run(
    profile: string,
    retries = 3,
  ): Promise<UserProfileSummarizationResult | null> {
    Logger.debug('Start summarizing user profile...')
    let retry = 0
    while (retry < retries) {
      try {
        const output = await this.chain.invoke({ profile })
        const startingIndex = output.indexOf('{')
        const jsonString = output.substring(startingIndex)
        const result = plainToInstance(
          UserProfileSummarizationResult,
          JSON.parse(jsonString),
        )

        await validateOrReject(result)
        return result
      } catch (e) {
        Logger.error(`Failed to summarize user profile: ${e}, retrying...`)
        retry += 1
      }
    }

    Logger.error(`Failed to summarize user profile after ${retry} retries.`)
    return null
  }
}
